wandb/sdk/launch/agent2/controllers/base.py

```diff
@@ -157,3 +193,146 @@ async def find_orphaned_jobs(self) -> List[Any]:
     @abstractmethod
     def label_job(self, project: LaunchProject) -> None:
         raise NotImplementedError
+
+    async def finish_launched_run(
+        self, run_with_tracker: RunWithTracker, status: Status
+    ) -> None:
+        run = run_with_tracker.run
+        tracker = run_with_tracker.tracker
+        item_id = tracker.run_queue_item_id
+        entity = tracker.entity
+        project = tracker.project
+        run_id = tracker.run_id
+
+        if entity is None or project is None or run_id is None:
+            self.logger.error(
+                f"called finish_thread_id on thread whose tracker has no project or run id. RunQueuetemID: {item_id}"
+            )
+
+            fail_run_queue_item = event_loop_thread_exec(
+                self.jobset.api.fail_run_queue_item
+            )
+            await fail_run_queue_item(
+                item_id,
+                "The submitted job was finished without assigned project or run id",
+                "agent",
+            )
+
+            return
+        run_called_init, logs = await check_run_called_init(
+            self.jobset.api, run, entity, project, run_id, item_id
+        )
+
+        if not run_called_init:
+            fnames = None
+            if logs:
+                fnames = tracker.saver.save_contents(logs, "error.log", "error")
+            if status == "finished":
+                _msg = "The submitted job exited successfully but failed to call wandb.init"
+            else:
+                _msg = f"The submitted job failed to call wandb.init exited with status: {status}"
+
+            fail_run_queue_item = event_loop_thread_exec(
+                self.jobset.api.fail_run_queue_item
+            )
+            await fail_run_queue_item(
+                item_id,
+                _msg,
+                "run",
+                fnames,
+            )
+
+    async def fail_run_with_exception(
+        self,
+        item_id: str,
+        exception: Exception,
+        tracker: Optional[JobAndRunStatusTracker] = None,
+    ) -> None:
+        tb_str = traceback.format_exception(
+            type(exception), value=exception, tb=exception.__traceback__
+        )
+        fnames = None
+        phase = None
+        if tracker is not None:
+            phase = tracker.err_stage
+            fnames = tracker.saver.save_contents("".join(tb_str), "error.log", "error")
+        else:
+            phase = "agent"
+
+        fail_run_queue_item = event_loop_thread_exec(
+            self.jobset.api.fail_run_queue_item
+        )
+        await fail_run_queue_item(item_id, str(exception), phase, fnames)
+
+        return
+
+    async def fail_unsubmitted_run(self, item_id: str) -> None:
```

Comment:
not specific to this line but IMO docstrings on all the BaseManager methods would go a long way, especially the abstract methods that subclasses are supposed to implement

Reply:
Fair! I'll add those